[WIP] Disable restarting and add wsgi callable

danlamanna opened the pull request:
@mgrauer @zachmullen This is the first PR related to running Girder on Beanstalk.

This adds a wsgi callable for Girder so running something like `uwsgi` is feasible:
`uwsgi --lazy --http :8080 --module girder.wsgi --check-static clients/web`

I'm unsure of how to handle the UI for disabling restart without rewriting it, frankly this would be easier if it were 2 buttons again. I'm going to be creating another PR which disables rebuilding web code (for when running on a CDN and/or having multiple app servers), so the single button would need to handle each scenario.

danlamanna commented:
@zachmullen If this looks good/makes sense to you I'll go ahead and add tests for the manage_plugins setting.

zachmullen commented on clients/web/src/views/body/PluginsView.js:
```diff
@@ -63,16 +64,30 @@ var PluginsView = View.extend({
                 this.render();
-            }, this));
+            }).fail(() => {
+                events.trigger('g:navigateTo', UsersView);
```
Why navigate to users view? That gets us in a weird state where the hash doesn't match the view, doesn't it?

danlamanna replied:
I was copying the logic present in the default routing for the plugins page.

zachmullen replied:
You've copied code written by [a fool](https://github.com/girder/girder/commit/014e26974c8f5ee7d580207bd3754af62e7b3ac1#diff-a7933f3a28741ea295cda1938cb28217R54)!

danlamanna replied:
Should I just `router.navigate` with trigger set to true? Or should it go somewhere else?

zachmullen replied:
Yeah, let's do that. And maybe we can go to the home page instead of #users.

zachmullen commented on girder/conf/girder.dist.cfg:
```diff
@@ -20,6 +20,10 @@ replica_set = None
 
+# Disable this if operating in a WSGI environment where Girder is unable
+# to restart itself
+restartable = True
```
Maybe we should make this setting more generic, like `cherrypy_server = True`. What do you think?

danlamanna replied:
Yeah that makes more sense, as long as people are familiar with the distinction between CherryPy's engine and server.

zachmullen replied:
We can just change the comment to make that distinction clear

danlamanna commented:
This is ready for review.

danlamanna commented:
Looking at it now @zachmullen, I suppose I should alter the plugins setting validator to check for this setting as well.

zachmullen commented:
We still need to be able to alter the plugins at some point or another, right?

danlamanna commented:
Yeah, my intention was just to disable the buttons. It seems weird that the
user can go over to the swagger page and circumvent that, but I suppose it
makes sense.

On Tue, Jun 6, 2017 at 10:58 AM, Zach Mullen <notifications@github.com>
wrote:

> We still need to be able to alter the plugins at some point or another,
> right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/girder/girder/pull/2038#issuecomment-306513102>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAlH5aiPqObLlmkCDV6BAXWviT4OTvbUks5sBWkAgaJpZM4Now20>
> .
>



-- 
Dan LaManna
Kitware, Inc.

brianhelba commented on clients/web/src/views/body/PluginsView.js:
```diff
@@ -63,16 +63,30 @@ var PluginsView = View.extend({
+                restRequest({
+                    path: 'system/plugins',
+                    type: 'GET'
+                }),
```
I'd tack on a
```
.then( (resp) => resp )
```
to the end of both of these `restRequest` calls here. That will transform the both promises to have only a single argument (dropping the extra `textStatus`, `jqXHR` arguments).

Then below, inside the `.done((plugins, managePlugins) => {` block, you can refer to the arguments directly, without the need for a confusing `[0]` index.

brianhelba commented on girder/wsgi.py:
```diff
@@ -0,0 +1,30 @@
+
+application = server.setup()
+
+cherrypy.server.unsubscribe()
```
[The docs](http://docs.cherrypy.org/en/latest/deploy.html#embedding-into-another-wsgi-framework) want you to maybe do `cherrypy.engine.signals.subscribe()`. Do you know what this is about, or why we don't need to do it here?

danlamanna replied:
It seems to be about determining who takes control of `SIGHUP` and `SIGTERM`. It also claims that it's dependent on which framework you're using. I'm opting to leave it out since their own uWSGI example (further down that page) does and I get a RuntimeWarning when adding it.

brianhelba commented on girder/wsgi.py:
```diff
@@ -0,0 +1,30 @@
+                        'environment': 'embedded'})
+cherrypy.config['server'].update({'cherrypy_server': False})
+
+application = server.setup()
```
Maybe add a comment here, noting that `application` is a magical variable name in WSGI?

brianhelba commented on girder/api/v1/system.py:
```diff
@@ -295,6 +295,9 @@ def discardPartialUploads(self, uploadId, userId, parentId, assetstoreId,
     )
     def restartServer(self, params):
+        if not config.getConfig()['server'].get('cherrypy_server', True):
+            raise RestException('Restarting of server is disabled.')
```
It's not really the client's fault if this operation is disallowed, so I don't think `400 Bad Request` is the appropriate error code. IMO, `403 Forbidden` is the most appropriate, since this is like:
* ([from the spec](https://tools.ietf.org/html/rfc7231#section-6.5.3)) `the server understood the request but refuses to authorize it` ... `a request might be forbidden for reasons unrelated to the credentials`
* [from the Wikipedia article](https://en.wikipedia.org/wiki/HTTP_403#Difference_from_status_.22401_Unauthorized.22) (which is admittedly less definitive) `The operation is forbidden to all users. For example, requests for a directory listing return code 403 when directory listing has been disabled.`

Of course, I don't think this is that critical, since Girder's error codes aren't always very compliant with HTTP.

brianhelba commented on clients/web/src/routes.js:
```diff
@@ -128,15 +128,7 @@ router.route('item/:id', 'item', function (itemId, params) {
-    }).done(_.bind(function (resp) {
-        events.trigger('g:navigateTo', PluginsView, resp);
-    }, this)).fail(_.bind(function () {
-        events.trigger('g:navigateTo', UsersView);
```
Do you have any idea why a fallback to `UsersView` was even here in the first place? It makes no sense to me.

danlamanna commented:
@brianhelba Thanks for the review, I've addressed some of your concerns but am taking a slightly different approach in this PR than when you reviewed it. I'll circle back and make sure all of your concerns that are still applicable are addressed.

@zachmullen Does the premise of this PR make a bit more sense now? Allow plugin enabling/disabling since that's valid, and just disable the restart button with a useful message when Girder isn't using CherryPy as the server?

brianhelba commented on tests/cases/system_test.py:
```diff
@@ -301,6 +301,15 @@ def testRestart(self):
+
+        resp = self.request(path='/system/restart', method='PUT',
+                            user=self.users[0])
+        self.assertStatus(resp, 400)
```
403 now.

danlamanna commented:
@manthey Can you sanity check/ack 6fe86bc?
